README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# twosamples

<!-- badges: start -->

[![R-CMD-check](https://github.com/cdowd/twosamples/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/cdowd/twosamples/actions/workflows/R-CMD-check.yaml)

[![pages-build-deployment](https://github.com/cdowd/twosamples/actions/workflows/pages/pages-build-deployment/badge.svg?branch=main)](https://github.com/cdowd/twosamples/actions/workflows/pages/pages-build-deployment)

<!-- badges: end -->

The goal of twosamples is to allow easy two-sample testing using
permutations. This package offers quick (O(n^2) worst) and memory light
(O(n)) code for a number of standard distance metrics between two
samples. Using permutations of observed samples to calculate null
distributions for those metrics, this package also provides assumption
free p-values for the samples coming from the same distribution.

Details about the DTS statistic’s calculation, and performance can be
found [here.](https://arxiv.org/abs/2007.01360)

Example R code which works to build test functions can be found on
github
[here.](https://github.com/cdowd/twosamples/blob/main/R/Extras/test_stats_R.R)
This code uses the release v1.0.0 algorithm for simplicity, and is in R
rather than C++, so is substantially slower.

## Installation

You can install the released version of twosamples from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("twosamples")
```

Or you can install the most recent (possibly unstable) version from
github:

``` r
library(remotes)
install_github("cdowd/twosamples")
```

#### System Requirements

Going forward (v2.0.0 onwards), `twosamples` depends on having C++11
installed. This is true of most R platforms and available for all of
them. Compilation also requires headers from the `cpp11` package.

## Example

In this example, we have 100 observations from two different Normal
distributions.

``` r
library(twosamples)
vec1 = rnorm(100)
vec2 = rnorm(100,0.5)
output = two_sample(vec1,vec2)
output
#> Test Stat   P-Value 
#>  11.33121   0.01500
summary(output)
#> DTS Test 
#> =========================
#> Test Statistic: 11.33121 
#>        P-Value: 0.015 *
#> - - - - - - - - - - - - -
#>      n1      n2 n.boots 
#>     100     100    2000 
#> =========================
#> Test stat rejection threshold for alpha = 0.05 is: 9.949129 
#> Null rejected: samples are from different distributions
plot(output)
```

<img src="man/figures/example1-1.png" width="100%" />

## Metric Example Calculations

This section will review each of the different tests offered. I’ll offer
a brief description of the test mathematically, then some quick
intuition. After that, I’ll show R code that would build the test
statistic – to provide a full description. That code is not the code
underlying this package, but a C++ rewrite of it is.

Each of these test statistics looks at the ECDFs which samples
correspond to. I’ll refer to E(x) as the ECDF of sample 1, F(x) as the
ECDF of sample 2, and G(x) as the ECDF of the joint sample. The ECDF of
course takes a real input and returns a value corresponding to the
portion of the sample which was observed less than or equal to that
point.

### Kolmogorov-Smirnov Test

The KS test finds the largest (vertical) difference between the two
ECDFs. See `ks_test()`.

<details>
<summary>
Demonstration R Code Calculating Test Statistic
</summary>

``` r
ks_stat_R = function(vec1,vec2,power=1) {
  n1 = length(vec1)
  n2 = length(vec2)
  n = n1+n2
  
  joint.sample = c(vec1,vec2)
  e = c(rep(1/n1,n1),rep(0,   n2))
  f = c(rep(0,   n1),rep(1/n2,n2))
  
  ind = order(joint.sample)
  d = joint.sample[ind]
  e = e[ind]
  f = f[ind]
  
  out = 0
  Ecur = 0
  Fcur = 0
  height = 0
  for (i in 1:(n-1)) {
    Ecur = Ecur + e[i]
    Fcur = Fcur + f[i]
    if (d[i] != d[i+1]) height = abs(Fcur-Ecur)
    if (height > out) out = height
  }
  out^power
}
```

</details>

In the example plot below, the KS statistic is the height of the
vertical black line.

![](man/figures/ks.png "Example KS stat plot")

### Kuiper Test

The Kuiper test is much the same as Kolmogorov-Smirnov, but it sums the
largest difference in each direction. i.e. it cares about the difference
between E(x)-F(x) and F(x)-E(x). In some sense this should be trading
some power against mean-shifts for power against variance changes. See
`kuiper_test()`

<details>
<summary>
Demonstration R Code Calculating Test Statistic
</summary>

``` r
kuiper_stat_R = function(vec1,vec2,power=1) { 
  n1 = length(vec1)
  n2 = length(vec2)
  n = n1+n2
  
  joint.sample = c(vec1,vec2)
  e = c(rep(1/n1,n1),rep(0,   n2))
  f = c(rep(0,   n1),rep(1/n2,n2))
  
  ind = order(joint.sample)
  d = joint.sample[ind]
  e = e[ind]
  f = f[ind]
  
  up = 0
  down = 0
  Ecur = 0
  Fcur = 0
  height = 0
  for (i in 1:(n-1)) {
    Ecur = Ecur + e[i]
    Fcur = Fcur + f[i]
    if (d[i] != d[i+1]) height = Fcur-Ecur
    if (height > up) up = height
    if (height < down) down = height
  }
  abs(down)^power + abs(up)^power
}
```

</details>

In the example plot below, the Kuiper statistic is the sum of the
heights of the vertical black lines.

![](man/figures/kuiper.png "Example Kuiper stat plot")

### Cramer-Von Mises

The Cramer-Von Mises criterion further extends the intuition of Kuiper
and KS. It is actually the full sum across every observation X of the
difference \|F(x)-E(x)\|. This use of the full joint sample gives it a
substantial power gain, particularly against higher moments shifting.
See `cvm_test()`.

<details>
<summary>
Demonstration R Code Calculating Test Statistic
</summary>

``` r
cvm_stat_R = function(vec1,vec2,power=2){
  n1 = length(vec1)
  n2 = length(vec2)
  n = n1+n2
  
  joint.sample = c(vec1,vec2)
  e = c(rep(1/n1,n1),rep(0,   n2))
  f = c(rep(0,   n1),rep(1/n2,n2))
  
  ind = order(joint.sample)
  d = joint.sample[ind]
  e = e[ind]
  f = f[ind]
  
  out = 0
  Ecur = 0
  Fcur = 0
  height = 0
  dups = 1
  for (i in 1:(n-1)) {
    Ecur = Ecur + e[i]
    Fcur = Fcur + f[i]
    height = abs(Fcur-Ecur)
    if (d[i] != d[i+1]) {
      out = out + (height^power)*dups
      dups = 1
    } else if (d[i] == d[i+1]) {
      dups = dups+1
    }
  }
  out
}
```

</details>

In the example plot below, the CVM statistic is the sum of the heights
of the vertical black lines.

![](man/figures/cvm.png "Example CVM stat plot")

### Anderson-Darling

Anderson-Darling test starts from the Cramer-Von Mises logic. However,
they note that some points on the joint ECDF are higher variance than
others. Because there is more noise in those observations, they should
receive a lower weight. More than that, we can even describe the
‘optimal’ weighting function – it is closely related to the joint ECDF -
G. See `ad_test()`

<details>
<summary>
Demonstration R Code Calculating Test Statistic
</summary>

``` r
ad_stat_R = function(vec1,vec2,power=2){
  n1 = length(vec1)
  n2 = length(vec2)
  n = n1+n2
  
  joint.sample = c(vec1,vec2)
  e = c(rep(1/n1,n1),rep(0,   n2))
  f = c(rep(0,   n1),rep(1/n2,n2))
  
  ind = order(joint.sample)
  d = joint.sample[ind]
  e = e[ind]
  f = f[ind]
  
  out = 0
  Ecur = 0
  Fcur = 0
  Gcur = 0
  height = 0
  dups = 1
  
  for (i in 1:(n-1)) {
    Ecur = Ecur + e[i]
    Fcur = Fcur + f[i]
    Gcur = Gcur+1/n
    sd = (2*Gcur*(1-Gcur)/n)**0.5
    height = abs(Fcur-Ecur)
    if (d[i] != d[i+1]) {
      out = out + ((height/sd)^power)*dups
      dups = 1
    } else if (d[i] == d[i+1]) {
      dups = dups+1
    }
  }
  out
}
```

</details>

In the example plot below, we see the variance of the joint ECDF over
the range of the data. It clearly peaks in the middle of the joint
sample.

![](man/figures/var.png "Plot of Variance of joint ECDF")

In the example plot below, the AD statistic is the weighted sum of the
heights of the vertical lines, where weights are represented by the
shading of the lines.

![](man/figures/ad.png "Example AD stat plot")

### Wasserstein

The Wasserstein distance is not normally a two-sample distance, but the
extension is very simple. In terms of the ECDFs it can actually be
described as the area between the ECDFs. Intuitively, the improvement
this offers over CVM is that it allows more extreme values to change our
conclusions. That is to say – KS, Kuiper, CVM, and AD have all operated
on ordinal variables, with no sense of the distance between different
rank order observations. Wasserstein (and below DTS) will require
interval data. By utilizing the extra information encoded when
information is from interval data, Wasserstein and DTS will improve on
CVM and AD. See `wass_test()`

<details>
<summary>
Demonstration R Code Calculating Test Statistic
</summary>

``` r
wass_stat_R = function(vec1,vec2,power=1) {
  n1 = length(vec1)
  n2 = length(vec2)
  n = n1+n2
  
  joint.sample = c(vec1,vec2)
  e = c(rep(1/n1,n1),rep(0,   n2))
  f = c(rep(0,   n1),rep(1/n2,n2))
  
  ind = order(joint.sample)
  d = joint.sample[ind]
  e = e[ind]
  f = f[ind]
  
  out = 0
  Ecur = 0
  Fcur = 0
  height = 0
  width = 0
  for (i in 1:(n-1)) {
    Ecur = Ecur + e[i]
    Fcur = Fcur + f[i]
    height = abs(Fcur-Ecur)
    width = d[i+1]-d[i]
    out = out + (height^power)*width
  }
  out
}
```

</details>

In the example plot below, the Wasserstein statistic is the shaded area
between the ECDFs.

![](man/figures/wass.png "Example Wasserstein stat plot")

### DTS/two_sample

If the Wasserstein metric improves on CVM by moving it into the realm of
interval data, then DTS improves on AD by doing the same. Alternately –
DTS offers the same improvement over Wasserstein that AD offers over
CVM. See `dts_test()` (AKA `two_sample()` ).

<details>
<summary>
Demonstration R Code Calculating Test Statistic
</summary>

``` r
dts_stat_R = function(vec1,vec2,power=1) {
  n1 = length(vec1)
  n2 = length(vec2)
  n = n1+n2
  
  joint.sample = c(vec1,vec2)
  e = c(rep(1/n1,n1),rep(0,   n2))
  f = c(rep(0,   n1),rep(1/n2,n2))
  
  ind = order(joint.sample)
  d = joint.sample[ind]
  e = e[ind]
  f = f[ind]
  
  out = 0
  Ecur = 0
  Fcur = 0
  Gcur = 0
  height = 0
  width = 0 
  for (i in 1:(n-1)) {
    Ecur = Ecur + e[i]
    Fcur = Fcur + f[i]
    Gcur = Gcur+1/n
    sd = (2*Gcur*(1-Gcur)/n)**0.5
    height = abs(Fcur-Ecur)
    width = d[i+1]-d[i]
    out = out + ((height/sd)^power)*width
  }
  out
}
```

</details>

In the example plot below, the DTS statistic is the shaded area between
the ECDFs, weighted by the variances – shown by the color of the
shading.

![](man/figures/dts.png "Example Wasserstein stat plot")

## Permutation Testing

Once we have a metric measuring distance between two samples, its easy
enough to code up a testing procedure.

Specifically we want to test the following: \* H0: J=K

where Sample 1 is distributed i.i.d. from J and Sample 2 is i.i.d. from
K.

Broadly, under this null, observations from sample 1 and sample 2 are
exchangeable.That is – by randomly swapping observations between the
samples, we can generate a new, equally likely data set from the same
(Null) DGP. Then we can calculate our metric for distance between the
samples on this data set. By repeating this procedure many times, we can
generate a null distribution for our distance metric. We can then
compare our observed distance to the null distribution. Because
distances are non-negative, this will be a one-sided comparison. (Though
if you wanted to test whether a sample was hyper-regular you could do
that by looking at the one-sided test to the bottom).

Moreover, unlike other, more general, two-sample procedures we do not
rely on any properties of the underlying distribution. Some procedures
require continuity, derivatives, or smoothness to get their asymptotic
and finite sample guarantees. Those are procedures which estimate the

NEWS.md
# version 2.0.1
This update contains two minor bugfixes. 

1. Some of the unit tests use random numbers, one of those unit tests looking at 
the function `combine.twosamples` failed  to consider default (desired) behavior 
for some possible values of those random numbers, and would occasionally fail. 
This was a minor issue that would not negatively impact any users, _except_ that
it would occasionally cause a CRAN check to fail, which could cause trouble. 
This update aims to fix the _test_, leaving all actual functionality unchanged, 
so that there should not be further CRAN check failures from this issue. 
2. Removed "C++11" from "System Requirements" -- as per current CRAN guidelines. 
The presence of C++11 as a requirement caused a note for R-CMD-Check in R4.3.0 and up.
Currently, the lack of such a requirement does not affect compilation on any tested systems. 

# version 2.0.0
This is a large update. Three big changes:

1. Substantial speed improvements -- should improve both absolute speed and scaling
2. Added generics for plot, summary, print.
3. Switched from the package `Rcpp` to `cpp11` for the backend. This removes a runtime dependency on `Rcpp`, but adds one on `C++11` and adds a compile time dependency on `cpp11`.

Together, all this lead to many changes under the hood. As a consequence, `permutation_test_builder` is substantially different (and no longer exported), and `order_stl` no longer exists. 

#### Speed
Each run of a *_test function now only sorts the data one time. Denoting the joint sample size N and the number of bootstraps K, this update moves the code from $O(KN\log(N))$ to $O(KN)+O(N \log(N))$. 

- Particularly for large samples or large numbers of bootstraps, this means a substantial improvement in speed. 
- This required reworking the underlying C++ `_stat` functions as well as the `permutation_test_builder`. 
  - Instead of breaking code in unpredictable ways, this function is no longer exported. If you used it, archived copies can be found on github (particularly under the example R code versions), or by emailing me. 
- Functions `*_stat` that are syntactically identical to the old ones still exist, but are no longer what is used by `permutation_test_builder`.
- These changes likely reduced memory requirements for most users, though this is offset by the new default of storing bootstrap outputs. 

#### Classes
There is now a 'twosamples' class, and generics for `print`, `summary`, and `plot`, as well as a function for combining outputs correctly. This should make the printed behavior much better. As well as making it easy to see a fair bit of information using summary. 

- plotting currently shows a histogram of the bootstrap values and a red line where the test-statistic is.
  - This required making the `*_test` functions export the bootstrap values. If you have memory intensive applications, this can be turned off with a toggle `keep.boots`, at the cost of no longer being able to use the plotting. 
  - In the future I may add the ability to plot the ECDFs and the test stat images. This is the main reason for the `keep.samples` toggle which is turned off by default.

#### Possible breaking changes
- In order to only sort once, this is now a proper permutation test again. This should also resolve some classes of potential validity issues. Proofs in the associated paper are (at the moment) not relevant to this for the same reason.
- `order_stl` no longer exists. I do not believe anybody used this function outside its internal package use. 
- `permutation_test_builder` is no longer exported. I am not aware of anyone using this function outside its internal package use. A similar function is still available, but will require changing the syntax of functions for its inputs.
- Dependency switch from `Rcpp` to `cpp11` and an additional system requirement of `C++11`. 


# version 1.2.0
This version is primarily bug fixes and documentation updates. **These bug fixes may affect outputs users see**. 

#### Summary

I expect this update to be purely cosmetic for the vast majority of users.
- For a few users of `ad_test` or `cvm_test` it is possible that re-running code will make significant differences to conclusions. 
- For the rare users of `ad_test` or `dts_test` (aka `two_sample`) who relied on the scale of the test stat (rather than merely the p-value), this update will change outputs substantially. In principal this change is merely re-scaling everything by $(n^p)/(2^{p/2})$. 


#### Main code changes:

- Fixed a major bug in how `ad_stat` and `cvm_stat` treated duplicates. This bug lead to excessive power in some situations. Re-running code, p-values and test stats may change.
- Fixed a minor bug in how `ad_stat` and `dts_stat` calculated standard deviations. Re-running code this will change the scale/location of the test stat, but should not affect p-values. 
- Some minor performance improvements: e.g. eliminated some unnecessary comparisons `if (sd >0)`. 
- renamed a functions internal variables to prevent an unlikely namespace conflict. 

#### Some documentation changes: 

- Website using pkgdown now exists at https://twosampletest.com
- link to website in description
- Fixed an error in the documentation describing `ad_stat` and `dts_stat` -- in which a square root term was dropped
- updated discussion of order_stl
- added some notes about ability to use factors (ordered or not)

#### Some development oriented changes:
- added some automated testing of the basic functions
- added reverse dependency testing
- added automated R-CMD-CHECK for each github commit


# version 1.1.1
This update is only fixing up documentation. Fixes a bug that lead to poor formatting, improves formatting of equations, adds graphs for test statistics, adds links between help pages. See v1.1.0 for recent improvements to codebase. 

# version 1.1.0

R/documentation.R
## usethis namespace: start
#' @useDynLib twosamples, .registration = TRUE
## usethis namespace: end
NULL

# KS test -----------

#' Kolmogorov-Smirnov Test
#'
#' @description A two-sample test using the Kolmogorov-Smirnov test statistic (`ks_stat`).
#' @param a a vector of numbers (or factors -- see details)
#' @param b a vector of numbers
#' @param nboots Number of bootstrap iterations
#' @param p power to raise test stat to
#' @param power power to raise test stat to
#' @param keep.boots Should the bootstrap values be saved in the output?
#' @param keep.samples Should the samples be saved in the output?
#' @return Output is a length 2 Vector with test stat and p-value in that order. That vector has 3 attributes -- the sample sizes of each sample, and the number of bootstraps performed for the pvalue.
#' @details The KS test compares two ECDFs by looking at the maximum difference between them. Formally -- if E is the ECDF of sample 1 and F is the ECDF of sample 2, then \deqn{KS = max |E(x)-F(x)|^p} for values of x in the joint sample. The test p-value is calculated by randomly resampling two samples of the same size using the combined sample.
#'
#' In the example plot below, the KS statistic is the height of the vertical black line.
#'
#' ![](ks.png "Example KS stat plot")
#'
#' Inputs `a` and `b` can also be vectors of ordered (or unordered) factors, so long as both have the same levels and orderings. When possible, ordering factors will substantially increase power.
#' @seealso [dts_test()] for a more powerful test statistic. See [kuiper_test()] or [cvm_test()] for the natural successors to this test statistic.
#' @examples
#' set.seed(314159)
#' vec1 = rnorm(20)
#' vec2 = rnorm(20,0.5)
#' out = ks_test(vec1,vec2)
#' out
#' summary(out)
#' plot(out)
#'
#' # Example using ordered factors
#' vec1 = factor(LETTERS[1:5],levels = LETTERS,ordered = TRUE)
#' vec2 = factor(LETTERS[c(1,2,2,2,4)],levels = LETTERS, ordered=TRUE)
#' ks_test(vec1,vec2)
#' @name ks_test
NULL

# Kuiper test ---------

#' Kuiper Test
#'
#' @description A two-sample test based on the Kuiper test statistic (`kuiper_stat`).
#' @param a a vector of numbers (or factors -- see details)
#' @param b a vector of numbers
#' @param nboots Number of bootstrap iterations
#' @param p power to raise test stat to
#' @param power power to raise test stat to
#' @param keep.boots Should the bootstrap values be saved in the output?
#' @param keep.samples Should the samples be saved in the output?
#' @return Output is a length 2 Vector with test stat and p-value in that order. That vector has 3 attributes -- the sample sizes of each sample, and the number of bootstraps performed for the pvalue.
#' @details The Kuiper test compares two ECDFs by looking at the maximum positive and negative difference between them. Formally -- if E is the ECDF of sample 1 and F is the ECDF of sample 2, then \deqn{KUIPER = |max_x E(x)-F(x)|^p + |max_x F(x)-E(x)|^p}. The test p-value is calculated by randomly resampling two samples of the same size using the combined sample.
#'
#' In the example plot below, the Kuiper statistic is the sum of the heights of the vertical black lines.
#'
#' ![](kuiper.png "Example Kuiper stat plot")
#'
#' Inputs `a` and `b` can also be vectors of ordered (or unordered) factors, so long as both have the same levels and orderings. When possible, ordering factors will substantially increase power.
#' @seealso [dts_test()] for a more powerful test statistic. See [ks_test()] for the predecessor to this test statistic, and [cvm_test()] for its natural successor.
#' @examples
#' set.seed(314159)
#' vec1 = rnorm(20)
#' vec2 = rnorm(20,0.5)
#' out = kuiper_test(vec1,vec2)
#' out
#' summary(out)
#' plot(out)
#'
#'
#' # Example using ordered factors
#' vec1 = factor(LETTERS[1:5],levels = LETTERS,ordered = TRUE)
#' vec2 = factor(LETTERS[c(1,2,2,2,4)],levels = LETTERS, ordered=TRUE)
#' kuiper_test(vec1,vec2)
#' @name kuiper_test
NULL

# CVM test ------------

#' Cramer-von Mises Test
#'
#' @description A two-sample test based on the Cramer-Von Mises test statistic (`cvm_stat`).
#' @param a a vector of numbers (or factors -- see details)
#' @param b a vector of numbers
#' @param nboots Number of bootstrap iterations
#' @param p power to raise test stat to
#' @param power power to raise test stat to
#' @param keep.boots Should the bootstrap values be saved in the output?
#' @param keep.samples Should the samples be saved in the output?
#' @return Output is a length 2 Vector with test stat and p-value in that order. That vector has 3 attributes -- the sample sizes of each sample, and the number of bootstraps performed for the pvalue.
#' @details The CVM test compares two ECDFs by looking at the sum of the squared differences between them -- evaluated at each point in the joint sample. Formally -- if E is the ECDF of sample 1 and F is the ECDF of sample 2, then \deqn{CVM = \sum_{x\in k}|E(x)-F(x)|^p}{CVM = SUM_(x in k) |E(x)-F(x)|^p} where k is the joint sample. The test p-value is calculated by randomly resampling two samples of the same size using the combined sample. Intuitively the CVM test improves on KS by using the full joint sample, rather than just the maximum distance -- this gives it greater power against shifts in higher moments, like variance changes.
#'
#' In the example plot below, the CVM statistic is the sum of the heights of the vertical black lines.
#'
#' ![](cvm.png "Example CVM stat plot")
#'
#' Inputs `a` and `b` can also be vectors of ordered (or unordered) factors, so long as both have the same levels and orderings. When possible, ordering factors will substantially increase power.
#' @seealso [dts_test()] for a more powerful test statistic. See [ks_test()] or [kuiper_test()] for the predecessors to this test statistic. See [wass_test()] and [ad_test()] for the successors to this test statistic.
#' @examples
#' set.seed(314159)
#' vec1 = rnorm(20)
#' vec2 = rnorm(20,0.5)
#' out = cvm_test(vec1,vec2)
#' out
#' summary(out)
#' plot(out)
#'
#' # Example using ordered factors
#' vec1 = factor(LETTERS[1:5],levels = LETTERS,ordered = TRUE)
#' vec2 = factor(LETTERS[c(1,2,2,2,4)],levels = LETTERS, ordered=TRUE)
#' cvm_test(vec1,vec2)
#' @name cvm_test
NULL


# AD test ------------------

#' Anderson-Darling Test
#'
#' @description A two-sample test based on the Anderson-Darling test statistic (`ad_stat`).
#' @param a a vector of numbers (or factors -- see details)
#' @param b a vector of numbers
#' @param nboots Number of bootstrap iterations
#' @param p power to raise test stat to
#' @param power power to raise test stat to
#' @param keep.boots Should the bootstrap values be saved in the output?
#' @param keep.samples Should the samples be saved in the output?
#' @return Output is a length 2 Vector with test stat and p-value in that order. That vector has 3 attributes -- the sample sizes of each sample, and the number of bootstraps performed for the pvalue.
#' @details The AD test compares two ECDFs by looking at the weighted sum of the squared differences between them -- evaluated at each point in the joint sample. The weights are determined by the variance of the joint ECDF at that point, which peaks in the middle of the joint distribution (see figure below). Formally -- if E is the ECDF of sample 1, F is the ECDF of sample 2, and G is the ECDF of the joint sample then \deqn{AD = \sum_{x \in k} \left({|E(x)-F(x)| \over \sqrt{2G(x)(1-G(x))/n} }\right)^p }{AD =  SUM_(x in k) (|E(x)-F(x)|/sqrt(2G(x)*(1-G(x)))/n)^p} where k is the joint sample. The test p-value is calculated by randomly resampling two samples of the same size using the combined sample. Intuitively the AD test improves on the CVM test by giving lower weight to noisy observations.
#'
#' In the example plot below, we see the variance of the joint ECDF over the range of the data. It clearly peaks in the middle of the joint sample.
#'
#' ![](var.png "Plot of Variance of joint ECDF")
#'
#' In the example plot below, the AD statistic is the weighted sum of the heights of the vertical lines, where weights are represented by the shading of the lines.
#'
#' ![](ad.png "Example AD stat plot")
#'
#' Inputs `a` and `b` can also be vectors of ordered (or unordered) factors, so long as both have the same levels and orderings. When possible, ordering factors will substantially increase power.
#' @seealso [dts_test()] for a more powerful test statistic. See [cvm_test()] for the predecessor to this test statistic. See [dts_test()] for the natural successor to this test statistic.
#' @examples
#' set.seed(314159)
#' vec1 = rnorm(20)
#' vec2 = rnorm(20,0.5)
#' out = ad_test(vec1,vec2)
#' out
#' summary(out)
#' plot(out)
#'
#' # Example using ordered factors
#' vec1 = factor(LETTERS[1:5],levels = LETTERS,ordered = TRUE)
#' vec2 = factor(LETTERS[c(1,2,2,2,4)],levels = LETTERS, ordered=TRUE)
#' ad_test(vec1,vec2)
#' @name ad_test
NULL


# Wass test ---------------

#' Wasserstein Distance Test
#' @description A two-sample test based on Wasserstein's distance (`wass_stat`).
#' @param a a vector of numbers (or factors -- see details)
#' @param b a vector of numbers
#' @param nboots Number of bootstrap iterations
#' @param p power to raise test stat to
#' @param power power to raise test stat to
#' @param keep.boots Should the bootstrap values be saved in the output?
#' @param keep.samples Should the samples be saved in the output?
#' @return Output is a length 2 Vector with test stat and p-value in that order. That vector has 3 attributes -- the sample sizes of each sample, and the number of bootstraps performed for the pvalue.
#' @details The Wasserstein test compares two ECDFs by looking at the Wasserstein distance between the two. This is of course the area between the two ECDFs. Formally -- if E is the ECDF of sample 1 and F is the ECDF of sample 2, then \deqn{WASS = \int_{x \in R} |E(x)-F(x)|^p}{WASS = Integral |E(x)-F(x)|^p} across all x. The test p-value is calculated by randomly resampling two samples of the same size using the combined sample. Intuitively the Wasserstein test improves on CVM by allowing more extreme observations to carry more weight. At a higher level -- CVM/AD/KS/etc only require ordinal data. Wasserstein gains its power because it takes advantages of the properties of interval data -- i.e. the distances have some meaning.
#'
#' In the example plot below, the Wasserstein statistic is the shaded area between the ECDFs.
#'
#' ![](wass.png "Example Wasserstein stat plot")
#'
#' Inputs `a` and `b` can also be vectors of ordered (or unordered) factors, so long as both have the same levels and orderings. When possible, ordering factors will substantially increase power. `wass_test` will assume the distance between adjacent factors is 1.
#' @seealso [dts_test()] for a more powerful test statistic. See [cvm_test()] for the predecessor to this test statistic. See [dts_test()] for the natural successor of this test statistic.
#' @examples
#' set.seed(314159)
#' vec1 = rnorm(20)
#' vec2 = rnorm(20,0.5)
#' out = wass_test(vec1,vec2)
#' out
#' summary(out)
#' plot(out)
#'
#' # Example using ordered factors
#' vec1 = factor(LETTERS[1:5],levels = LETTERS,ordered = TRUE)
#' vec2 = factor(LETTERS[c(1,2,2,2,4)],levels = LETTERS, ordered=TRUE)
#' wass_test(vec1,vec2)
#' @name wass_test
NULL


# DTS test ----------------------

#' DTS Test
#'
#' @description A two-sample test based on the DTS test statistic (`dts_stat`). This is the recommended two-sample test in this package because of its power. The DTS statistic is the reweighted integral of the distance between the two ECDFs.
#' @param a a vector of numbers (or factors -- see details)
#' @param b a vector of numbers
#' @param nboots Number of bootstrap iterations
#' @param p power to raise test stat to
#' @param power also the power to raise the test stat to
#' @param keep.boots Should the bootstrap values be saved in the output?
#' @param keep.samples Should the samples be saved in the output?
#' @return Output is a length 2 Vector with test stat and p-value in that order. That vector has 3 attributes -- the sample sizes of each sample, and the number of bootstraps performed for the pvalue.
#' @details The DTS test compares two ECDFs by looking at the reweighted Wasserstein distance between the two. See the companion paper at [arXiv:2007.01360](https://arxiv.org/abs/2007.01360) or <https://codowd.com/public/DTS.pdf> for details of this test statistic, and non-standard uses of the package (parallel for big N, weighted observations, one sample tests, etc).
#'
#' If the [wass_test()] extends [cvm_test()] to interval data, then [dts_test()] extends [ad_test()] to interval data. Formally -- if E is the ECDF of sample 1, F is the ECDF of sample 2, and G is the ECDF of the combined sample, then \deqn{DTS = \int_{x\in R} \left({|E(x)-F(x)| \over \sqrt{2G(x)(1-G(x))/n}}\right)^p}{DTS =  Integral (|E(x)-F(x)|/sqrt(2G(x)(1-G(x))/n))^p} for all x.
#' The test p-value is calculated by randomly resampling two samples of the same size using the combined sample. Intuitively the DTS test improves on the AD test by allowing more extreme observations to carry more weight. At a higher level -- CVM/AD/KS/etc only require ordinal data. DTS (and Wasserstein) gain power because they take advantages of the properties of interval data -- i.e. the distances have some meaning. However, DTS, like Anderson-Darling (AD) also downweights noisier observations relative to Wass, thus (hopefully) giving it extra power.
#'
#' In the example plot below, the DTS statistic is the shaded area between the ECDFs, weighted by the variances -- shown by the color of the shading.
#'
#' ![](dts.png "Example Wasserstein stat plot")
#'
#' Inputs `a` and `b` can also be vectors of ordered (or unordered) factors, so long as both have the same levels and orderings. When possible, ordering factors will substantially increase power. The dts test will assume the distance between adjacent factors is 1.
#' @seealso [wass_test()], [ad_test()] for the predecessors of this test statistic. [arXiv:2007.01360](https://arxiv.org/abs/2007.01360) or <https://codowd.com/public/DTS.pdf> for details of this test statistic
#' @examples
#' set.seed(314159)
#' vec1 = rnorm(20)
#' vec2 = rnorm(20,0.5)
#' dts_stat(vec1,vec2)
#' out = dts_test(vec1,vec2)
#' out
#' summary(out)
#' plot(out)
#' two_sample(vec1,vec2)
#'
#' # Example using ordered factors
#' vec1 = factor(LETTERS[1:5],levels = LETTERS,ordered = TRUE)
#' vec2 = factor(LETTERS[c(1,2,2,2,4)],levels = LETTERS, ordered=TRUE)
#' dts_test(vec1,vec2)
#' @name two_sample
NULL


# Perm Builder ---------------------

#' Permutation Test Builder
#'
#' @description (**Warning!** This function has changed substantially between v1.2.0 and v2.0.0) This function takes a two-sample test statistic and produces a function which performs randomization tests (sampling with replacement) using that test stat. This is an internal function of the `twosamples` package.
#' @param test_stat_function a function of the joint vector and a label vector producing a positive number, intended as the test-statistic to be used.
#' @param default.p This allows for some introduction of defaults and parameters. Typically used to control the power functions raise something to.
#' @return  This function returns a function which will perform permutation tests on given test stat.
#' @details test_stat_function must be structured to take two vectors -- the first a combined sample vector and the second a logical vector indicating which sample each value came from, as well as a third and fourth value. i.e. (fun = function(jointvec,labelvec,val1,val2) ...). See examples.
#'
#' ### Conversion Function
#' Test stat functions designed to work with the prior version of `permutation_test_builder` will not work.
#' E.g. If your test statistic is
#' ```
#' mean_diff_stat = function(x,y,pow) abs(mean(x)-mean(y))^pow
#' ```
#' then `permutation_test_builder(mean_diff_stat,1)` will no longer work as intended, but it will if you run the below code first.
#' ```
#' perm_stat_helper = function(stat_fn,def_power) {
#'   output = function(joint,vec_labels,power=def_power,na) {
#'     a = joint[vec_labels]
#'     b = joint[!vec_labels]
#'     stat_fn(a,b,power)
#'   }
#'   output
#' }
#'
#' mean_diff_stat = perm_stat_helper(mean_diff_stat)
#' ```
#' @examples
#' mean_stat = function(joint,label,p,na) abs(mean(joint[label])-mean(joint[!label]))**p
#' myfun = twosamples:::permutation_test_builder(mean_stat,2.0)
#' set.seed(314159)
#' vec1 = rnorm(20)
#' vec2 = rnorm(20,0.5)
#' out = myfun(vec1,vec2)
#' out
#' summary(out)
#' plot(out)
#'
#' @name permutation_test_builder
#' @seealso [two_sample()]
NULL

# twosamples-class ---------------------

#' twosamples_class
#'
#' @description Objects of Class twosamples are output by all of the `*_test` functions in the `twosamples` package.
#'
#' @details
#' By default they consist of:a length 2 vector, the first item being the test statistic, the second the p-value.
#' That vector has the following attributes:
#'
#' 1. details: length 3 vector with the sample sizes for each sample and the number of bootstraps
#' 2. test_type: a string describing the type of the test statistic
#'
#' It may also have two more attributes, depending on options used when running the `*_test` function. These are useful for plotting and combining test runs.
#'
#' 1. bootstraps: a vector containing all the bootstrapped null values
#' 2. samples: a list containing both the samples that were tested
#'
#' and by virtue of being a named length 2 vector of class "twosamples" it has the following two attributes:
#' 1. names: c("Test Stat","P-Value")
#' 2. class: "twosamples"
#'
#' Multiple Twosamples objects made by the same `*_test` routine being run on the same data can be combined (getting correct p-value and correct attributes) with the function `combine_twosamples()`.
#'
#' @param x twosamples object
#' @param object twosamples-object to summarize
#' @param alpha  Significance threshold for determining null rejection
#' @param ... other parameters to be passed to print or summary functions
#'
#' @returns
#' - `print.twosamples()` returns nothing
#' - `summarize.twosamples()` returns nothing
#' @seealso [plot.twosamples()], [combine.twosamples()]
#' @name twosamples_class
NULL
